### PR TITLE
REST API for Users Improvements

### PR DESCRIPTION
**Pull request by vboctor**
- Implemented UserGetCommand
- [27128](https://www.mantisbt.org/bugs/view.php?id=27128): Can not get userid from another user with REST API - Added `GET /users/username/:username` to get a user by name and `/users/:id` .
- [27133](https://www.mantisbt.org/bugs/view.php?id=27133) delete user by username - approach: get user info by username, then delete by id.
- [32357](https://www.mantisbt.org/bugs/view.php?id=32357): REST API: Support select for fields to return when getting user info - this applies to `/users/me`, `/users/:id`, `/users/username/:username` and include select query string with field names.

### CODE REVIEW

**Comment by vboctor** [comment]
I'm not sure why a test is failing on Travis, it is passing on my machine and it is checking for a user that was created by the test case. I'll look into it more, strange!

File: core/commands/UserGetCommand.php [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+				}
+			}
+		} else {
+			$this->target_user_id = (int)$t_user_id;
+		}
+
+		if( !is_null( $t_user_ref ) && $this->target_user_id == 0 ) {
+			throw new ClientException(
+				"User not found with id, name, realname or email equal to '$t_user_ref'",
+				ERROR_USER_BY_NAME_NOT_FOUND,
```
Comment (dregad) [inline core/commands/UserGetCommand.php:106]: We actually have distinct error codes for the various scenarios (ERROR_USER_BY_ID_NOT_FOUND, ERROR_USER_BY_EMAIL_NOT_FOUND, ERROR_USER_BY_REALNAME_NOT_FOUND), so the Exception should use them and not throw a potentially incorrect ERROR_USER_BY_NAME_NOT_FOUND.

Set a variable with the error code in the if blocks above.

File: api/rest/restcore/users_rest.php [hunk 2]
```diff
@@ -76,20 +106,37 @@ function rest_user_get_me( \Slim\Http\Request $p_request, \Slim\Http\Response $p
...
+		$t_select = array(
+			'id',
+			'name',
+			'real_name',
+			'email',
+			'access_level',
+			'language',
+			'timezone',
+			'created_at'
+		);
```
Comment (dregad) [inline api/rest/restcore/users_rest.php:122]: The Command already defines the default fields, is it necessary to duplicate code here ?

File: core/commands/UserGetCommand.php [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+				$this->select = $t_select;
+			}
+		}
+
+		# if select field is not specified or empty, then use defaults
+		if( is_null( $this->select ) ) {
+			$this->select = array(
+				'id',
+				'name',
+				'real_name',
```
Comment (dregad) [inline core/commands/UserGetCommand.php:142]: What happens if user has no access to real_name ?

**Review by dregad — CHANGES_REQUESTED** [review]
Email and Real Name have no guarantee to be unique. If there are multiple matches, the API will just return the first one (exactly which one is not defined) and caller will never know that there are other occurrences.

Therefore I don't think it's a good idea to allow querying by email or real name using user_get_id_by_XXX as-is. IMO, in this scenario we should bet a list of matching users.

Also consider you introduced user_get_id_by_user_info() some years back, so there is some duplicated logic.

File: core/commands/UserGetCommand.php (hunk 1 above)
Comment (vboctor) [reply core/commands/UserGetCommand.php:106]: I have changed the API to be:

/users/me
/users/:user_id
/users/username/:username

And remove by email and real name.

File: api/rest/restcore/users_rest.php (hunk 2 above)
Comment (vboctor) [reply api/rest/restcore/users_rest.php:122]: I have refactored the code to have a method that returns the default fields and used it from everywhere. The /users/me uses default fields and adds `projects` field to the list.

File: core/commands/UserGetCommand.php (hunk 3 above)
Comment (vboctor) [reply core/commands/UserGetCommand.php:142]: If user doesn't have access to real name, then they won't see it. If they are retrieving their own user, then they will see it. That is a standard behavior for getting account ref structure across the board.

**Pull request merged**